src/huggingface_hub/repocard.py

```diff
@@ -99,3 +104,156 @@ def metadata_eval_result(
+
+    for key in metadata:
+        # update all fields except model index
+        if key != "model-index":
+            if key in existing_metadata and not overwrite:
+                if existing_metadata[key] != metadata[key]:
+                    raise ValueError(
+                        f"""You passed a new value for the existing meta data field '{key}'. Set `overwrite=True` to overwrite existing metadata."""
+                    )
+            else:
+                existing_metadata[key] = metadata[key]
+        # update model index containing the evaluation results
+        else:
+            if "model-index" not in existing_metadata:
+                existing_metadata["model-index"] = metadata["model-index"]
+            else:
+                # the model-index contains a list of results as used by PwC but only has one element thus we take the first one
+                existing_metadata["model-index"][0][
+                    "results"
+                ] = _update_metadata_model_index(
+                    existing_metadata["model-index"][0]["results"],
+                    metadata["model-index"][0]["results"],
+                    overwrite=overwrite,
+                )
+
+    # save and push to hub
+    metadata_save(filepath, existing_metadata)
+
+    HfApi().upload_file(
+        path_or_fileobj=filepath,
+        path_in_repo="README.md",
+        repo_id=repo_id,
+        repo_type=repo_type,
+        identical_ok=True,
+        token=token,
+    )
+
+
+def _update_metadata_model_index(existing_results, new_results, overwrite=False):
+    """
+    Updates the model-index fields in the metadata. If results with same unique
+    features exist they are updated, else a new result is appended. Updating existing
+    values is only possible if `overwrite=True`.
+
+    Args:
+        new_metrics (`list`):
+            List of new metadata results
+        existing_metrics (`list`):
+            List of existing metadata results
+        overwrite (`bool`, *optional*, defaults to `False`):
+            If set to `True` an existing metric values can be overwritten, otherwise
+            attempting to overwrite an existing field will cause an error.
+
+    Returns:
+        `list`: List of updated metadata results
+    """
+    for new_result in new_results:
+        result_found = False
+        for existing_result_index, existing_result in enumerate(existing_results):
+            if all(
+                new_result[feat] == existing_result[feat]
+                for feat in UNIQUE_RESULT_FEATURES
+            ):
+                result_found = True
+                existing_results[existing_result_index][
+                    "metrics"
+                ] = _update_metadata_results_metric(
+                    new_result["metrics"],
+                    existing_result["metrics"],
+                    overwrite=overwrite,
+                )
+        if not result_found:
+            existing_results.append(new_result)
```

Comment:
Since dictionaries are mutable, you could also write this as:

```py
        try:
            existing_result = next(
                x
                for x in existing_results
                if all(x[feat] == new_result[feat] for feat in UNIQUE_RESULT_FEATURES)
            )
            existing_result["metrics"] = _update_metadata_results_metric(
                new_result["metrics"],
                existing_result["metrics"],
                overwrite=overwrite,
            )
        except StopIteration:
            existing_results.append(new_result)
```

which should be faster since it avoids one slow for loop.